### PR TITLE
Add integration/system tests for Kafka JMX metricsets

### PR DESCRIPTION
**Pull request by ChrsMark**
This PR adds integration and system tests for Kafka JMX metricsets as a follow-up of https://github.com/elastic/beats/pull/14330

### CODE REVIEW

File: metricbeat/tests/system/test_kafka.py [hunk 1]
```diff
@@ -46,6 +46,83 @@ def test_partition(self):
 
         self.assert_fields_are_documented(evt)
 
+    @unittest.skipUnless(metricbeat.INTEGRATION_TESTS, "integration test")
+    def test_consumer(self):
```
Comment (jsoriano) [inline metricbeat/tests/system/test_kafka.py]: If these three tests are equal consider using `parameterized`, take a look to `test_ceph.py` or `test_couchbase.py` for  examples.

If you need multiple parameters in principle you can also do it, e.g:
```
@parameterized.expand([
  ('consumer', '8774/tcp'),
  ('producer', '8775/tcp'),
  ('broker', '8779/tcp'),
])
def test_kafka(self, metricset, port):
    ...
```

File: metricbeat/module/kafka/broker/_meta/fields.yml [hunk 2]
```diff
@@ -37,14 +43,26 @@
...
+      description: The outgoing byte rate per topic
       type: float
     - name: topic.net.bytes_rejected
-      description: The rejected byte rate
+      description: The rejected byte rate per topic
       type: float
     - name: topic.messages_in
+      description: The incoming message rate per topic
+      type: float
+    - name: net.bytes_in
```
Comment (jsoriano) [inline metricbeat/module/kafka/broker/_meta/fields.yml]: Following field conventions this should probably be:
```suggestion
    - name: net.in.bytes_per_sec
```

File: metricbeat/module/kafka/broker/_meta/fields.yml [hunk 3]
```diff
@@ -37,14 +43,26 @@
...
+      description: The incoming byte rate
+      type: float
+    - name: net.bytes_out
```
Comment (jsoriano) [inline metricbeat/module/kafka/broker/_meta/fields.yml]: And:
```suggestion
    - name: net.out.bytes_per_sec
```

File: metricbeat/module/kafka/broker/_meta/fields.yml [hunk 4]
```diff
@@ -37,14 +43,26 @@
...
+      description: The outgoing byte rate
+      type: float
+    - name: net.bytes_rejected
```
Comment (jsoriano) [inline metricbeat/module/kafka/broker/_meta/fields.yml]: And:
```suggestion
    - name: net.rejected.bytes_per_sec
```

File: metricbeat/module/kafka/broker/broker_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func TestFetch(t *testing.T) {
+	service := compose.EnsureUp(t, "kafka",
+		compose.UpWithTimeout(600*time.Second),
+		compose.UpWithAdvertisedHostEnvFileForPort(9092),
+	)
+	reporter := &mbtest.CapturingReporterV2{}
+
+	metricSet := mbtest.NewReportingMetricSetV2Error(t, getConfig(service.HostForPort(8779)))
```
Comment (jsoriano) [inline metricbeat/module/kafka/broker/broker_integration_test.go]: Please use the generic methods:
```
metricSet := mbtest.NewEventsFetcher(t, getConfig(service.HostForPort(8779)))
events, err := metricSet.Fetch()
require.NoError(t, err)
...
```
You can find some examples in other modules.

The advantage is that if at some moment the interface of the fetcher changes, the test doesn't need to be changed.

File: metricbeat/module/kafka/broker/broker_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	service := compose.EnsureUp(t, "kafka",
+		compose.UpWithTimeout(600*time.Second),
+		compose.UpWithAdvertisedHostEnvFileForPort(9092),
+	)
+
+	ms := mbtest.NewReportingMetricSetV2Error(t, getConfig(service.HostForPort(8779)))
+	err := mbtest.WriteEventsReporterV2Error(ms, t, "")
+	if err != nil {
+		t.Fatal("write", err)
+	}
```
Comment (jsoriano) [inline metricbeat/module/kafka/broker/broker_integration_test.go]: Also here use the generic methods:
```
        f := mbtest.NewEventsFetcher(t, getConfig(service.HostForPort(8779))
        err := mbtest.WriteEvents(f, t)
        require.NoError(t, err)
```

File: metricbeat/tests/system/test_kafka.py [hunk 7]
```diff
@@ -36,7 +37,36 @@ def test_partition(self):
...
+        evt = output[0]
+        print(evt)
+
+        self.assert_fields_are_documented(evt)
+
+    @parameterized.expand([
+        ('consumer', '8774/tcp'),
+        ('producer', '8775/tcp'),
+        ('broker', '8779/tcp'),
+    ])
```
Comment (jsoriano) [inline metricbeat/tests/system/test_kafka.py]: :+1:

File: metricbeat/module/kafka/consumer/consumer_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func TestData(t *testing.T) {
+	service := compose.EnsureUp(t, "kafka",
+		compose.UpWithTimeout(600*time.Second),
+		compose.UpWithAdvertisedHostEnvFileForPort(9092),
+	)
+
+	ms := mbtest.NewReportingMetricSetV2Error(t, getConfig(service.HostForPort(8774)))
+	err := mbtest.WriteEventsReporterV2Error(ms, t, "")
```
Comment (jsoriano) [inline metricbeat/module/kafka/consumer/consumer_integration_test.go]: Also use generic methods in this test file.

File: metricbeat/module/kafka/producer/producer_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+func TestFetch(t *testing.T) {
+	service := compose.EnsureUp(t, "kafka",
+		compose.UpWithTimeout(600*time.Second),
+		compose.UpWithAdvertisedHostEnvFileForPort(9092),
+	)
+	reporter := &mbtest.CapturingReporterV2{}
+
+	metricSet := mbtest.NewReportingMetricSetV2Error(t, getConfig(service.HostForPort(8775)))
+	metricSet.Fetch(reporter)
```
Comment (jsoriano) [inline metricbeat/module/kafka/producer/producer_integration_test.go]: And also use generic test methods in this file.

File: metricbeat/module/kafka/broker/_meta/fields.yml (hunk 2 above)
Comment (ChrsMark) [reply metricbeat/module/kafka/broker/_meta/fields.yml]: @sorantis is this ok from product perspective or we should strictly follow the naming provided in the initial [issue](https://github.com/elastic/beats/issues/13366#issue-486253665)?

Comment (sorantis) [reply metricbeat/module/kafka/broker/_meta/fields.yml]: there's no strict requirement on the naming. The field names were taken from the ER attached to the original issue. Do we have the mentioned naming convention somewhere?

Comment (jsoriano) [reply metricbeat/module/kafka/broker/_meta/fields.yml]: You can find the conventions here: https://www.elastic.co/guide/en/beats/devguide/7.5/event-conventions.html

**Comment by ChrsMark** [comment]
@jsoriano thanks for the suggestions! Changed the fields' naming since @sorantis went along and implemented the tests using the `NewFetcher`.

**Pull request merged**